cmd/influx/main.go

```diff
@@ -185,26 +186,72 @@ func newApiClient(ctx *cli.Context, cli *internal.CLI, injectToken bool) (*api.A
 	return api.NewAPIClient(apiConfig), nil
 }
 
-// standardCtx returns a context that will cancel on SIGINT and SIGTERM.
-func standardCtx(ctx *cli.Context) context.Context {
-	return signals.WithStandardSignals(ctx.Context)
-}
-
 var app = cli.App{
 	Name:      "influx",
 	Usage:     "Influx Client",
 	UsageText: "influx [command]",
 	Commands: []*cli.Command{
 		&versionCmd,
 		&pingCmd,
-		&setupCmd,
-		&writeCmd,
-		&bucketCmd,
+		newSetupCmd(),
+		newWriteCmd(),
+		newBucketCmd(),
 	},
 }
 
+func withCli() cli.BeforeFunc {
+	return func(ctx *cli.Context) error {
+		c, err := newCli(ctx)
+		if err != nil {
+			return err
+		}
+		ctx.App.Metadata["cli"] = c
+		return nil
+	}
+}
+
+func getCLI(ctx *cli.Context) *internal.CLI {
+	i, ok := ctx.App.Metadata["cli"].(*internal.CLI)
+	if !ok {
+		panic("missing CLI")
+	}
+	return i
+}
+
+func withApi() cli.BeforeFunc {
+	makeFn := func(key string, injectToken bool) cli.BeforeFunc {
+		return func(ctx *cli.Context) error {
+			c := getCLI(ctx)
+			apiClient, err := newApiClient(ctx, c, injectToken)
+			if err != nil {
+				return err
+			}
+			ctx.App.Metadata[key] = apiClient
+			return nil
+		}
+	}
+	return middleware.WithBeforeFns(makeFn("api", true), makeFn("api-no-token", false))
+}
+
+func getAPI(ctx *cli.Context) *api.APIClient {
+	i, ok := ctx.App.Metadata["api"].(*api.APIClient)
+	if !ok {
+		panic("missing APIClient with token")
+	}
+	return i
+}
+
+func getAPINoToken(ctx *cli.Context) *api.APIClient {
+	i, ok := ctx.App.Metadata["api-no-token"].(*api.APIClient)
+	if !ok {
+		panic("missing APIClient without token")
+	}
+	return i
+}
+
 func main() {
-	if err := app.Run(os.Args); err != nil {
+	ctx := signals.WithStandardSignals(context.Background())
+	if err := app.RunContext(ctx, os.Args); err != nil {
```

Comment:
We run the app with a context that contains the standard signals, saving us from having to do that for every sub command